From 803be375c9676ae4c8f9399872cb9cc3539cb70a Mon Sep 17 00:00:00 2001
From: Darren Salt <linux@youmustbejoking.demon.co.uk>
Date: Sun, 28 Jan 2018 18:20:19 +0900
Subject: [PATCH 02/10] escape '^From ' lines for not breaking GPG MIME
 signatures

---
 libsylph/quoted-printable.c |    7 +++++++
 1 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/libsylph/quoted-printable.c b/libsylph/quoted-printable.c
index adcdb05..adee4bc 100644
--- a/libsylph/quoted-printable.c
+++ b/libsylph/quoted-printable.c
@@ -52,6 +52,13 @@ void qp_encode_line(gchar *out, const guchar *in)
 			if (*inp == '\r')
 				inp++;
 			inp++;
+			if (strncmp(inp, "From ", 5) == 0) {
+				*outp++ = '=';
+				*outp++ = '4';
+				*outp++ = '6';
+				len += 3;
+				inp++;
+			}
 		} else if (ch == '\t' || ch == ' ') {
 			if (IS_LBREAK(inp + 1)) {
 				SOFT_LBREAK_IF_REQUIRED(3);
-- 
1.7.9

