From e82b5967c1cc4de00cd9e0160895470373e4b55f Mon Sep 17 00:00:00 2001
From: Peter Zhigalov <peter.zhigalov@gmail.com>
Date: Mon, 28 Nov 2022 23:05:37 +0700
Subject: [PATCH 07/10] Update default font for Windows

---
 libsylph/defs.h |    6 +++++-
 1 files changed, 5 insertions(+), 1 deletions(-)

diff --git a/libsylph/defs.h b/libsylph/defs.h
index 18f0c97..10e6c9f 100644
--- a/libsylph/defs.h
+++ b/libsylph/defs.h
@@ -129,6 +129,10 @@
 #define SESSION_TIMEOUT_INTERVAL	60	/* sec */
 #define MAX_HISTORY_SIZE		16
 
-#define DEFAULT_MESSAGE_FONT	"Monospace 12"
+#ifdef G_OS_WIN32
+#  define DEFAULT_MESSAGE_FONT	"Courier New 10"
+#else
+#  define DEFAULT_MESSAGE_FONT	"Monospace 12"
+#endif
 
 #endif /* __DEFS_H__ */
-- 
1.7.9

