From d1cce9d2e37a9b27b1c932d8b258c75684e98153 Mon Sep 17 00:00:00 2001
From: Peter Zhigalov <peter.zhigalov@gmail.com>
Date: Mon, 5 Dec 2022 02:15:53 -0800
Subject: [PATCH 2/5] localedir-relocation

---
 gtkspell/gtkspell.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/gtkspell/gtkspell.c b/gtkspell/gtkspell.c
index 1f70b05..40a8f3b 100644
--- a/gtkspell/gtkspell.c
+++ b/gtkspell/gtkspell.c
@@ -740,7 +740,13 @@ gtkspell_new_attach(GtkTextView *view, const gchar *lang, GError **error) {
 	GtkSpell *spell;
 
 #ifdef ENABLE_NLS
+#ifdef G_OS_WIN32
+	gchar *dir = g_win32_get_package_installation_directory_of_module(NULL);
+	bindtextdomain(PACKAGE, g_build_filename (dir, "share", "locale", NULL));
+	g_free(dir);
+#else
 	bindtextdomain(PACKAGE, LOCALEDIR);
+#endif
 	bind_textdomain_codeset(PACKAGE, "UTF-8");
 #endif
 
-- 
2.38.1.windows.1

